pkg/adaptor/cloud/vsphere/provider.go

```diff
@@ -122,9 +124,114 @@ func (p *vsphereProvider) CreateInstance(ctx context.Context, podName, sandboxID
 
 	vmfolderref := vmfolder.Reference()
 
-	relocateSpec := types.VirtualMachineRelocateSpec{
-		Folder: &vmfolderref,
-		Pool:   poolref,
+	var relocateSpec types.VirtualMachineRelocateSpec
+
+	relocateSpec.Folder = &vmfolderref
+
+	cloneSpec := &types.VirtualMachineCloneSpec{
+		PowerOn:  true,
+		Template: false,
+	}
+
+	var poolref *types.ManagedObjectReference
+
+	if strings.EqualFold(p.serviceConfig.DRS, "true") {
+
+		// If the user has selected DRS we check that there is a cluster name,
+		// ignore any non relevant inputs, and the rest is up to DRS
+
+		if p.serviceConfig.Cluster == "" {
+			err = fmt.Errorf("DRS requires a cluster name")
+			return nil, err
+		}
+
+		logger.Printf("Looking for cluster %s DRS recommendations", p.serviceConfig.Cluster)
+
+		cluster, err := finder.ClusterComputeResourceOrDefault(ctx, p.serviceConfig.Cluster)
+		if err != nil {
+			logger.Printf("Cluster %s compute resource error: %s", p.serviceConfig.Cluster, err)
+			return nil, err
+		}
+
+		vmref := vm.Reference()
+		spec := types.PlacementSpec{
+			PlacementType: string(types.PlacementSpecPlacementTypeClone),
+			CloneName:     vmname,
+			CloneSpec:     cloneSpec,
+			RelocateSpec:  &relocateSpec,
+			Vm:            &vmref,
+		}
+
+		result, err := cluster.PlaceVm(ctx, spec)
+		if err != nil {
+			logger.Printf("Cluster %s placement error: %s", p.serviceConfig.Cluster, err)
+			return nil, err
+		}
+
+		recs := result.Recommendations
+		if len(recs) == 0 {
+			return nil, fmt.Errorf("Vcenter has no cluster recommendations for cluster %s", p.serviceConfig.Cluster)
+		}
+
+		rspec := *recs[0].Action[0].(*types.PlacementAction).RelocateSpec
+		relocateSpec.Datastore = rspec.Datastore
+		relocateSpec.Host = rspec.Host
+		relocateSpec.Pool = rspec.Pool
+
+	} else if p.serviceConfig.Host != "" {
+
+		// Here a cluster name is required if there is a cluster and no DRS. The host hangs off the cluster
```

Comment:
Can you please add a reference doc link which describes the DRS flow or just describe the algo ? This will help to understand the entire logical flow for anyone looking at this later on

Reply:
I actually have this in the second patch but I can include it here in this one.

Reply:
I can describe it better.

Reply:
@bpradipt OK I think I addressed your concerns. I also decided to remove the call to  DefaultResoucePool. This was intended to be a last ditch effort to identify a resource pool when the user had not provided enough information. Unfortunately it only works in very simple config e.g. a datacenter with only one cluster with one host and one datastore. So it basically allows the user to skirt following the config rules which can cause confusion if any additions to the datacenter are made and the call no longer works.